File: SIMPLEAPI_Demo/SIMPLEAPI_Demo/Impresion/Core/Classes/PrinterRow.cs
using SIMPLEAPI_Demo.Impresion.Core.Helpers;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SIMPLEAPI_Demo.Impresion.Core.Classes
{
    public class PrinterRow
    {
        public enum TextOverflowResolveMethods
        {
            Truncate = 0,
            Wrap = 1
        }

        public TextOverflowResolveMethods TextOverflowResolveMethod { get; set; }

        public Color Color { get; set; }
        public int FontSize { get; set; }
        public bool IsBold { get; set; }

        public List<PrinterColumn> Columns { get; private set; }

        public PrinterRow()
        {
            Columns = new List<PrinterColumn>();
        }

        public void AddColumn(PrinterColumn col)
        {
            Columns.Add(col);
        }

        public void ValidateColumnsWidth(bool testing = false)
        {
            if (testing)
                return;

            var total_chars = Columns.Sum(x => x.MaxChars);
            var max_chars_font = FontInfo.GetMaxChars(FontSize, IsBold ? FontStyle.Bold : FontStyle.Regular);
            if (total_chars > max_chars_font)
                throw new Exception("Max chars for font currently in use must not be over " + max_chars_font);
        }

        public List<string> ToLines()
        {
            List<string> Lines = new List<string>();
            var text = "";
            var aux = "";
            int i, j;

            for (int colIndex = 0; colIndex < Columns.Count; colIndex++)
            {
                var column = Columns[colIndex];

                if (column is PrinterTextColumn)
                {
                    var col = column as PrinterTextColumn;

                    text = col.Text ?? "";
                    switch (col.Align)
                    {
                        case PrinterColumn.Aligns.Center:
                            text = text.CenterTo(col.MaxChars);
                            break;
                        case PrinterColumn.Aligns.Left:
                            break;
                        case PrinterColumn.Aligns.Right:
                            text = text.PadLeft(col.MaxChars, ' ');
                            break;
                    }

                    if (TextOverflowResolveMethod == TextOverflowResolveMethods.Truncate)
                        text = text.Truncate(col.MaxChars);

                    for (i = 0, j = 0; i < text.Length; i += col.MaxChars, j++)
                    {
                        if (Lines.Count == j)
                            Lines.Add("");

                        // Iniciar nueva línea respetando el espacio de las columnas anteriores
                        if (i != 0 && Lines.Count - 1 == j && colIndex > 0)
                            Lines[j] += "".PadLeft(Columns.Take(colIndex - 1).Sum(x => x.MaxChars) + 1, ' ');

                        // Extraer porción del texto que cabe en una línea
                        aux = text.Substring(i, i + col.MaxChars > text.Length ? text.Length - i : col.MaxChars).PadRight(col.MaxChars, ' ');

                        // La línea no puede iniciar con un espacio, por lo tanto, se corren hasta el final.
                        if (aux.Length != aux.TrimStart().Length)
                            aux = aux.TrimStart() + "".PadLeft(aux.Length - aux.TrimStart().Length, ' ');

                        // respetar alineación de la columna en nueva línea
                        switch (col.Align)
                        {
                            case PrinterColumn.Aligns.Center:
                                aux = aux.CenterTo(col.MaxChars);
                                break;
                            case PrinterColumn.Aligns.Left:
                                break;
                            case PrinterColumn.Aligns.Right:
                                aux = aux.Trim().PadLeft(col.MaxChars, ' ');
                                break;
                        }

                        Lines[j] += aux;
                    }
                    if (j < Lines.Count)
                        for (i = j; i < Lines.Count; i++)
                            Lines[i] += "".PadRight(col.MaxChars, ' ');
                }
                else if (column is PrinterImageColumn)
                {
                    var col = column as PrinterImageColumn;
                    Lines.Add($"<img>{ col.Image }</img>");
                }
                else
                    throw new Exception("col type not supported");
            }
            return Lines;
        }
    }
}

File: SIMPLEAPI_Demo/SIMPLEAPI_Demo_VB/Impresion/Core/Classes/PrinterRow.vb



Public Class PrinterRow
    Public Enum TextOverflowResolveMethods
        Truncate = 0
        Wrap = 1
    End Enum

    Public Property TextOverflowResolveMethod As TextOverflowResolveMethods
    Public Property Color As Color
    Public Property FontSize As Integer
    Public Property IsBold As Boolean
    Public Property Columns As List(Of PrinterColumn)

    Public Sub New()
        Columns = New List(Of PrinterColumn)()
    End Sub

    Public Sub AddColumn(ByVal col As PrinterColumn)
        Columns.Add(col)
    End Sub

    Public Sub ValidateColumnsWidth(ByVal Optional testing As Boolean = False)
        If testing Then Return
        Dim total_chars = Columns.Sum(Function(x) x.MaxChars)
        Dim max_chars_font = FontInfo.GetMaxChars(FontSize, If(IsBold, FontStyle.Bold, FontStyle.Regular))
        If total_chars > max_chars_font Then Throw New Exception("Max chars for font currently in use must not be over " & max_chars_font)
    End Sub

    Public Function ToLines() As List(Of String)
        Dim Lines As List(Of String) = New List(Of String)()
        Dim text = ""
        Dim aux = ""
        Dim i, j As Integer

        For colIndex As Integer = 0 To Columns.Count - 1
            Dim column = Columns(colIndex)

            If TypeOf column Is PrinterTextColumn Then
                Dim col = TryCast(column, PrinterTextColumn)
                text = If(col.Text, "")

                Select Case col.Align
                    Case PrinterColumn.Aligns.Center
                        text = text.CenterTo(col.MaxChars)
                    Case PrinterColumn.Aligns.Left
                    Case PrinterColumn.Aligns.Right
                        text = text.PadLeft(col.MaxChars, " "c)
                End Select

                If TextOverflowResolveMethod = TextOverflowResolveMethods.Truncate Then text = text.Truncate(col.MaxChars)
                i = 0
                j = 0

                While i < text.Length
                    If Lines.Count = j Then Lines.Add("")
                    ' // Iniciar nueva línea respetando el espacio de las columnas anteriores
                    If i <> 0 AndAlso Lines.Count - 1 = j AndAlso colIndex > 0 Then Lines(j) += "".PadLeft(Columns.Take(colIndex - 1).Sum(Function(x) x.MaxChars) + 1, " "c)
                    '// Extraer porción del texto que cabe en una línea

                    aux = text.Substring(i, If(i + col.MaxChars > text.Length, text.Length - i, col.MaxChars)).PadRight(col.MaxChars, " "c)

                    '// La línea no puede iniciar con un espacio, por lo tanto, se corren hasta el final.

                    If aux.Length <> aux.TrimStart().Length Then aux = aux.TrimStart() & "".PadLeft(aux.Length - aux.TrimStart().Length, " "c)


                    '// respetar alineación de la columna en nueva línea

                    Select Case col.Align
                        Case PrinterColumn.Aligns.Center
                            aux = aux.CenterTo(col.MaxChars)
                        Case PrinterColumn.Aligns.Left
                        Case PrinterColumn.Aligns.Right
                            aux = aux.Trim().PadLeft(col.MaxChars, " "c)
                    End Select

                    Lines(j) += aux
                    i += col.MaxChars
                    j += 1
                End While

                If j < Lines.Count Then

                    For i = j To Lines.Count - 1
                        Lines(i) += "".PadRight(col.MaxChars, " "c)
                    Next
                End If
            ElseIf TypeOf column Is PrinterImageColumn Then
                Dim col = TryCast(column, PrinterImageColumn)
                Lines.Add($"<img>{col.Image}</img>")
            Else
                Throw New Exception("col type not supported")
            End If
        Next

        Return Lines
    End Function
End Class

